tests/sparseml/pytorch/sparsification/distillation/test_per_layer_distillation.py

```diff
@@ -0,0 +1,410 @@
+        project_features=False,
+        epsilon=3.0,
+    )
+    yaml_str = f"""
+        !PerLayerDistillationModifier
+            gain: {truth.gain}
+            start_epoch: {truth.start_epoch}
+            end_epoch: {truth.end_epoch}
+            update_frequency: {truth.update_frequency}
+            normalize: {truth.normalize}
+            student_layer_names: {truth.student_layer_names}
+            teacher_layer_names: {truth.teacher_layer_names}
+            project_features: {truth.project_features}
+            epsilon: {truth.epsilon}
+        """
+    from_yaml = PerLayerDistillationModifier.load_obj(yaml_str)
+    twice_from_yaml = PerLayerDistillationModifier.load_obj(str(from_yaml))
+
+    assert isinstance(from_yaml, PerLayerDistillationModifier)
+    for attr_name in [
+        "gain",
+        "start_epoch",
+        "end_epoch",
+        "update_frequency",
+        "normalize",
+        "student_layer_names",
+        "teacher_layer_names",
+        "project_features",
+        "epsilon",
+    ]:
+        assert (
+            getattr(truth, attr_name)
+            == getattr(from_yaml, attr_name)
+            == getattr(twice_from_yaml, attr_name)
+        )
+
+
+@pytest.mark.skipif(
+    os.getenv("NM_ML_SKIP_PYTORCH_TESTS", False),
+    reason="Skipping pytorch tests",
+)
+def test_optimizer_serialization_with_projections(tmp_path):
+    # since we are adding param groups to optimizer during initialization, we need
+    # to ensure that they can be serialized/reloaded properly
+    optim_checkpoint_path = tmp_path / "optim.pth"
+    student = mlp(12, 24, 64)
+    teacher = mlp(12, 24, 64)
+
+    modifier = PerLayerDistillationModifier()
+    modifier.initialize(module=student, distillation_teacher=teacher)
+
+    optimizer = create_optim_sgd(student)
+    x = torch.randn(5, 12)
+    fake_loss = student(x).mean()
+
+    # add state to optimizer
+    modifier.loss_update(
+        loss=fake_loss,
+        module=student,
+        optimizer=optimizer,
+        epoch=modifier.start_epoch,
+        steps_per_epoch=10,
+        student_inputs=x,
+        student_outputs=fake_loss,
+    )
+
+    optimizer_checkpoint = {
+        "optimizer": optimizer.state_dict(),
+    }
+    torch.save(optimizer_checkpoint, optim_checkpoint_path)
+    loaded_optimizer_checkpoint = torch.load(optim_checkpoint_path)
+
+    loaded_optimizer = create_optim_sgd(student)
+    loaded_optimizer.load_state_dict(loaded_optimizer_checkpoint["optimizer"])
+
+
+@pytest.mark.skipif(
+    os.getenv("NM_ML_SKIP_PYTORCH_TESTS", False),
+    reason="Skipping pytorch tests",
+)
+def test_hooks_survive_after_quantization_is_applied():
+    # this should test fused module hooks are properly kept around during
+    # quantization
+    modifier = PerLayerDistillationModifier()
+    student = resnet18(pretrained=True)
+    teacher = resnet18(pretrained=True)
+    opt = create_optim_sgd(student)
+
+    modifier.initialize(student, distillation_teacher=teacher)
+
```

Comment:
let's run a fake forward pass here before the quantization modifier is added as well for completeness since additional logic may happen here

Reply:
Done!